File: frontend/lib/backend/monitor.dart
class Monitor {
  ApiUsage? apiUsage;
  String? compiler;
  int? numGoroutine;
  int? cpuCore;
  BuildInfo? buildInfo;
  Memory? swapMemory;
  Memory? memory;
  int? totalBook;
  List<CpuStats>? cpuStats;
  List<double>? cpuPercents;

  Monitor(
      {this.apiUsage,
      this.compiler,
      this.numGoroutine,
      this.cpuCore,
      this.buildInfo,
      this.swapMemory,
      this.memory,
      this.totalBook,
      this.cpuStats,
      this.cpuPercents});

  Monitor.fromJson(Map<String, dynamic> json) {
    apiUsage =
        json['api_usage'] != null ? ApiUsage.fromJson(json['api_usage']) : null;
    compiler = json['compiler'];
    numGoroutine = json['num_goroutine'];
    cpuCore = json['cpu_core'];
    buildInfo = json['build_info'] != null
        ? BuildInfo.fromJson(json['build_info'])
        : null;
    swapMemory = json['swap_memory'] != null
        ? Memory.fromJson(json['swap_memory'])
        : null;
    memory = json['memory'] != null ? Memory.fromJson(json['memory']) : null;
    totalBook = json['total_book'];
    if (json['cpu_stats'] != null) {
      cpuStats = <CpuStats>[];
      json['cpu_stats'].forEach((v) {
        cpuStats!.add(CpuStats.fromJson(v));
      });
    }

    cpuPercents ??= <double>[];

    for (var percent in json["cpu_percents"]) {
      cpuPercents!.add((percent is double)
          ? percent
          : (percent is int)
              ? percent.toDouble()
              : 0);
    }
    //cpuPercents = json['cpu_percents'].cast<double>();
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {};
    if (apiUsage != null) {
      data['api_usage'] = apiUsage!.toJson();
    }
    data['compiler'] = compiler;
    data['num_goroutine'] = numGoroutine;
    data['cpu_core'] = cpuCore;
    if (buildInfo != null) {
      data['build_info'] = buildInfo!.toJson();
    }
    if (swapMemory != null) {
      data['swap_memory'] = swapMemory!.toJson();
    }
    if (memory != null) {
      data['memory'] = memory!.toJson();
    }
    data['total_book'] = totalBook;
    if (cpuStats != null) {
      data['cpu_stats'] = cpuStats!.map((v) => v.toJson()).toList();
    }
    data['cpu_percents'] = cpuPercents;
    return data;
  }
}

class ApiUsage {
  int? alloc;
  int? totalAlloc;
  int? numGc;

  ApiUsage({this.alloc, this.totalAlloc});

  ApiUsage.fromJson(Map<String, dynamic> json) {
    alloc = json['alloc'];
    totalAlloc = json['total_alloc'];
    numGc = json["num_gc"];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {};
    data['alloc'] = alloc;
    data['total_alloc'] = totalAlloc;
    data["num_gc"] = numGc;
    return data;
  }
}

class BuildInfo {
  String? goVersion;
  String? path;

  List<Settings>? settings;

  BuildInfo({this.goVersion, this.path, this.settings});

  BuildInfo.fromJson(Map<String, dynamic> json) {
    goVersion = json['GoVersion'];
    path = json['Path'];

    if (json['Settings'] != null) {
      settings = <Settings>[];
      json['Settings'].forEach((v) {
        settings!.add(Settings.fromJson(v));
      });
    }
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {};
    data['GoVersion'] = goVersion;
    data['Path'] = path;

    if (settings != null) {
      data['Settings'] = settings!.map((v) => v.toJson()).toList();
    }
    return data;
  }
}

class Settings {
  String? key;
  String? value;

  Settings({this.key, this.value});

  Settings.fromJson(Map<String, dynamic> json) {
    key = json['Key'];
    value = json['Value'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {};
    data['Key'] = key;
    data['Value'] = value;
    return data;
  }
}

// class SwapMemory {
//   int? total;
//   int? used;
//   int? free;
//   int? usedPercent;

//   SwapMemory({this.total, this.used, this.free, this.usedPercent});

//   SwapMemory.fromJson(Map<String, dynamic> json) {
//     total = json['Total'];
//     used = json['Used'];
//     free = json['Free'];
//     usedPercent = json['UsedPercent'];
//   }

//   Map<String, dynamic> toJson() {
//     final Map<String, dynamic> data = {};
//     data['Total'] = total;
//     data['Used'] = used;
//     data['Free'] = free;
//     data['UsedPercent'] = usedPercent;
//     return data;
//   }
// }

class Memory {
  int? total;
  int? used;
  int? free;
  double? usedPercent;

  Memory({this.total, this.used, this.free, this.usedPercent});

  Memory.fromJson(Map<String, dynamic> json) {
    total = json['Total'];
    used = json['Used'];
    free = json['Free'];
    usedPercent = (json['UsedPercent'] is double)
        ? json['UsedPercent']
        : json['UsedPercent'].toDouble();
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {};
    data['Total'] = total;
    data['Used'] = used;
    data['Free'] = free;
    data['UsedPercent'] = usedPercent ?? 0;
    return data;
  }
}

class CpuStats {
  int? cpu;
  String? vendorId;
  String? family;
  String? model;
  int? stepping;
  String? physicalId;
  String? coreId;
  int? cores;
  String? modelName;
  double? mhz;
  int? cacheSize;
  List<String>? flags;
  String? microcode;

  CpuStats(
      {this.cpu,
      this.vendorId,
      this.family,
      this.model,
      this.stepping,
      this.physicalId,
      this.coreId,
      this.cores,
      this.modelName,
      this.mhz,
      this.cacheSize,
      this.flags,
      this.microcode});

  CpuStats.fromJson(Map<String, dynamic> json) {
    cpu = json['cpu'];
    vendorId = json['vendorId'];
    family = json['family'];
    model = json['model'];
    stepping = json['stepping'];
    physicalId = json['physicalId'];
    coreId = json['coreId'];
    cores = json['cores'];
    modelName = json['modelName'];

    mhz = (json["mhz"] is double)
        ? json["mhz"]
        : (json["mhz"] is int ? json["mhz"].toDouble() : 0);
    cacheSize = json['cacheSize'];
    flags = json['flags'].cast<String>();
    microcode = json['microcode'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {};
    data['cpu'] = cpu;
    data['vendorId'] = vendorId;
    data['family'] = family;
    data['model'] = model;
    data['stepping'] = stepping;
    data['physicalId'] = physicalId;
    data['coreId'] = coreId;
    data['cores'] = cores;
    data['modelName'] = modelName;
    data['mhz'] = mhz;
    data['cacheSize'] = cacheSize;
    data['flags'] = flags;
    data['microcode'] = microcode;
    return data;
  }
}

File: backend/libsohal/monitor.go
package libsohal

import (
	"encoding/json"
	"runtime"
	"runtime/debug"
	"time"

	"github.com/gofiber/fiber/v2"
	"github.com/shirou/gopsutil/v3/cpu"
	"github.com/shirou/gopsutil/v3/mem"
)

type DetailedMemoryUsage struct {
	Alloc      uint64 `json:"alloc"`
	TotalAlloc uint64 `json:"total_alloc"`
	NumGc      int    `json:"num_gc"`
}

type Memoryusage struct {
	Total       uint64
	Used        uint64
	Free        uint64
	UsedPercent float64
}

type monitorstruct struct {
	DetailedMemoryUsage DetailedMemoryUsage `json:"api_usage"`
	Compiler            string              `json:"compiler"`
	NumGoroutine        int                 `json:"num_goroutine"`
	ServerCore          int                 `json:"cpu_core"`
	BuildInfo           debug.BuildInfo     `json:"build_info"`
	SwapMemory          Memoryusage         `json:"swap_memory"`
	Memory              Memoryusage         `json:"memory"`
	//TotalBook           int64               `json:"total_book"`
	CpuStats    []cpu.InfoStat `json:"cpu_stats"`
	CpuPercents []float64      `json:"cpu_percents"`
}

func (m monitorstruct) ToJson() []byte {
	j, _ := json.Marshal(m)
	return j
}

func (l *LibSohal) MonitorRouter(c *fiber.Ctx) error {
	var most monitorstruct = monitorstruct{
		Compiler:     runtime.Compiler,
		ServerCore:   runtime.NumCPU(),
		NumGoroutine: runtime.NumGoroutine(),
	}
	var _apimem runtime.MemStats
	runtime.ReadMemStats(&_apimem)

	most.DetailedMemoryUsage = DetailedMemoryUsage{
		Alloc:      _apimem.Alloc,
		TotalAlloc: _apimem.TotalAlloc,
		NumGc:      int(_apimem.NumGC),

		//Heap: _apimem,
	}

	//fmt.Printf("(_apimem.Alloc / 1024 / 1024): %v\n", (_apimem.Alloc / 1024 / 1024))
	// var rmem syscall.Rusage
	// if err := syscall.Getrusage(syscall.RUSAGE_SELF, &rmem); err != nil {
	// 	log.Fatalln(err)
	// }

	// fmt.Printf("(rmem.Maxrss / 1024 / 1024): %v\n", (rmem.Maxrss / 1024 / 1024))

	// if err := syscall.Getrusage(syscall.RUSAGE_CHILDREN, &rmem); err != nil {
	// 	log.Fatalln(err)
	// }

	// fmt.Printf("child (rmem.Maxrss / 1024 / 1024): %v\n", (rmem.Maxrss / 1024 / 1024))

	// if err := syscall.Getrusage(syscall.RUSAGE_THREAD, &rmem); err != nil {
	// 	log.Fatalln(err)
	// }

	// fmt.Printf("thread (rmem.Maxrss / 1024 / 1024): %v\n", (rmem.Maxrss / 1024 / 1024))

	buildinfo, ok := debug.ReadBuildInfo()
	if !ok {
		return fiber.NewError(500, "Cloudn't read GoBuildInfo")
	}
	most.BuildInfo = *buildinfo

	swapmem, err := mem.SwapMemoryWithContext(c.Context())
	if err != nil {
		return fiber.NewError(500, err.Error())
	}
	most.SwapMemory = Memoryusage{
		Total:       swapmem.Total,
		Used:        swapmem.Used,
		Free:        swapmem.Total - swapmem.Used,
		UsedPercent: swapmem.UsedPercent,
	}

	mem, err := mem.VirtualMemoryWithContext(c.Context())
	if err != nil {
		return fiber.NewError(500, err.Error())
	}

	most.Memory = Memoryusage{
		Total:       mem.Total,
		Used:        mem.Used,
		Free:        mem.Total - mem.Used,
		UsedPercent: mem.UsedPercent,
	}

	stats, err := cpu.InfoWithContext(c.Context())
	if err != nil {
		return fiber.NewError(500, err.Error())
	}
	most.CpuStats = stats

	percents, err := cpu.PercentWithContext(c.Context(), time.Second, true)
	if err != nil {
		return fiber.NewError(500, err.Error())
	}
	most.CpuPercents = percents

	// counter, err := l.Mongo.Database(template.MongoBookDatabase).Collection(template.MongoBookCollection).CountDocuments(c.Context(), map[any]any{})
	// if err != nil {
	// 	return fiber.NewError(500, err.Error())
	// }
	// most.TotalBook = counter

	return c.Send(most.ToJson())

}
